libs/ml/include/ml/graph.hpp

```diff
@@ -282,5 +301,65 @@ void Graph<ArrayType>::LoadStateDict(struct fetch::ml::StateDict<ArrayType> cons
   }
 }
 
+/**
+ * Assigns all trainable weights parameters to vector of ArrayType for exporting and serialising
+ * @return ret is vector containing values for all weights
+ */
+template <typename ArrayType>
+std::vector<ArrayType> Graph<ArrayType>::GetWeights() const
+{
+  std::vector<ArrayType> ret;
+
+  for (auto const &t : trainable_)
+  {
+    ret.push_back(t.second->GetWeights());
+  }
+  return ret;
+}
+
+/**
+ * Assigns all trainable accumulated gradient parameters to vector of ArrayType for exporting and
+ * serialising
+ * @return ret is vector containing all gradient values
+ */
+template <typename ArrayType>
+std::vector<ArrayType> Graph<ArrayType>::GetGradients() const
+{
+  std::vector<ArrayType> ret;
+
+  for (auto const &t : trainable_)
+  {
+    ret.push_back(t.second->Gradients());
+  }
+  return ret;
+}
+
+/**
+ * Sets all accumulated gradients for each trainable to zero
+ */
+template <typename ArrayType>
+void Graph<ArrayType>::ResetGradients()
+{
+  for (auto const &t : trainable_)
+  {
+    t.second->ResetGradients();
+  }
+}
+
+/**
+ * Add gradient values to weight for each trainable
+ * @param grad vector of gradient values for each trainable stored in ArrayType
+ */
+template <typename ArrayType>
+void Graph<ArrayType>::ApplyGradients(std::vector<ArrayType> &grad)
+{
+  typename ArrayType::SizeType i = 0;
```

Comment:
here you're mixing an iterator with indexing. probably better to use two iterators: i.e.
auto grad_it = grad.begin();
for(auto const &t : trainable_)
{
  t.second->ApplyGradient(*grad_it);
  ++grad_it;
}